src/tools/opensc-explorer.c

```diff
@@ -1670,28 +1763,31 @@ static int do_get_data(int argc, char **argv)
 static int do_put_data(int argc, char **argv)
 {
 	unsigned int tag;
-	u8 buf[8192];
-	size_t buflen = sizeof(buf);
+	u8 buf[SC_MAX_EXT_APDU_BUFFER_SIZE];
+	size_t i, len;
 	int r;
 
-	if (argc != 2)
+	if (argc < 2)
 		return usage(do_put_data);
 
 	/* Extract DO's tag */
 	tag = strtoul(argv[0], NULL, 16);
 
 	/* Extract the new content */
-	/* buflen is the max length of reception buffer */
-	r = parse_string_or_hexdata(argv[1], buf, &buflen);
-	if (r < 0) {
-		printf("unable to parse data\n");
-		return -1;
+	for (i = 1, len = 0; i < (unsigned) argc; i++)   {
+		size_t len0 = sizeof(buf) - len;
```

Comment:
Length checking is missing completely. I think it's possible to have an integer underflow here.

Reply:
Oops, correct finding.

In current opensc-explorer it cannot happen because `sizeof(buf)` is larger than the whole command line (and thus also larger than the sum of the parsed command line arguments).
But we should not rely on that.

Let me see what I can do.